README.md

```diff
@@ -1,8 +1,161 @@
-[![Build Status](https://travis-ci.org/tgu/kb_uploadmethods.svg?branch=master)](https://travis-ci.org/tgu/kb_uploadmethods)
-
 # kb_uploadmethods
----
 
-Uploaders for kbase datatypes, especially for files in stageing area
+[![Build Status](https://travis-ci.org/tgu/kb_uploadmethods.svg?branch=master)](https://travis-ci.org/tgu/kb_uploadmethods)  
+[Release Notes](RELEASE_NOTES.md)
+
+## Description
+This module implements [KBase](https://www.kbase.us) apps that are used to transform data files into KBase data objects for use in analysis. This also includes a few apps for adding external files to the user's [staging area](https://docs.kbase.us/getting-started/narrative/add-data#uploading-data-from-external-sources), and manipulating them there.
+
+See [here](https://docs.kbase.us/data/upload-download-guide/uploads) for more information on uploading data to KBase.
+
+## Development
+This module was created using the KBase SDK. See the [documentation](https://kbase.github.io/kb_sdk_docs/) for more detail. Apps are mostly broken down into submodules.
+
+The main entrypoint is in `lib/kb_uploadmethods/kb_uploadmethodsImpl.py`. Individual apps in that module use one or more utility module to handle different file types. These are all under `lib/kb_uploadmethods/Utils`. Add a new Util module if you're adding an uploader for a new data type.
+
+## Testing
+This module can be tested with the following steps, common to all KBase SDK modules.
+1. Install the [KBase SDK](https://kbase.github.io/kb_sdk_docs/).
+2. Fetch this module and navigate to it from the console.
+3. Run `kb-sdk test` once - this will initialize the `test_local/test.cfg` file.
+4. Populate the `test_local/test.cfg` file with a KBase authentication token.
+5. Run `kb-sdk test` again.
+
+Once steps 1-4 have been run once, you can just run `kb-sdk test` to run the test suite any time.
+
+## Apps
+
+Each app is listed below with the following format:
+**Display name** - what the user sees - links to module doc page, if available
+**app id**: id - how the app is identified to the system, including the directory under ui/narrative/methods  
+**entrypoint**: function - name of the function that gets run in this module, as described in `kb_uploadmethods.spec`  
```

Comment:
Are there any required input parameters for each app?

Reply:
@eamahanna yes, they are located in the specfile. https://github.com/kbaseapps/kb_uploadmethods/blob/master/kb_uploadmethods.spec

Reply:
There are input requirements for each app. I thought about adding them, but it made me concerned about expanding the document a little too much. I'll add a few and see if it gets too crazy. I can also add a Table of Contents pointing to each app for information.

Reply:
I'm worried about the burden of having to update the docs 3 times:
1) the comments or function signature of the code
2) the comments or function signature for the spec file
3) the comments or function signature for this doc

Reply:
Well, the first is done automatically when you update the spec file and rebuild - and if you're modifying the function signature, then there's gonna be changes to deal with in the code anyway.

For the third, I think that updating the code to change around how functions are called should require some changes to the user/dev- facing documentation. Whether that's this file or another doesn't really change that, IMO. 

Also, to be more precise, I'm not documenting the function calls here, but the apps and how they're invoked in the Narrative interface - which can be very different from just calling the uploader function.

Reply:
For the first case I would bring up the example specific example of using the Utils pattern, for which type hints and documentation are maintained manually, by the developer.

impl.py

```
def upload_fasta(string path, map parameters)
"""
AUTOGENERATED DOCUMENTATION
"""
upload_success = utils.upload(path,parameters)
return upload_success
```


utils.py
```
def upload_fasta(path: str, parameters : dict[str])
"""
ref : a filepath for your fasta file in your staging area
parameters: organism_name, sequencer_type, etc
"""
...
```

I like the idea of documenting stuff, especially in a way that non-devs can read it and understand what is going on and what the code is trying to achieve/what it's steps are. Since this will be user facing documentation then we should definite keep it updated and maybe provide a link to it in the app itself then!